borg/repository.py

```diff
@@ -683,18 +735,32 @@ def _read(self, fd, fmt, header, segment, offset, acceptable_tags):
             raise IntegrityError('Invalid segment entry size [segment {}, offset {}]'.format(
                 segment, offset))
         length = size - fmt.size
-        data = fd.read(length)
-        if len(data) != length:
-            raise IntegrityError('Segment entry data short read [segment {}, offset {}]: expected {}, got {} bytes'.format(
-                segment, offset, length, len(data)))
-        if crc32(data, crc32(memoryview(header)[4:])) & 0xffffffff != crc:
-            raise IntegrityError('Segment entry checksum mismatch [segment {}, offset {}]'.format(
-                segment, offset))
+        if read_data:
+            data = fd.read(length)
+            if len(data) != length:
+                raise IntegrityError('Segment entry data short read [segment {}, offset {}]: expected {}, got {} bytes'.format(
+                    segment, offset, length, len(data)))
+            if crc32(data, crc32(memoryview(header)[4:])) & 0xffffffff != crc:
+                raise IntegrityError('Segment entry checksum mismatch [segment {}, offset {}]'.format(
+                    segment, offset))
+            if key is None and tag in (TAG_PUT, TAG_DELETE):
+                key, data = data[:32], data[32:]
+        else:
```

Comment:
Yes. Meaning that we need to consider how corrupted data will propagate from here; this is used in two instances:

1. Reading length of to-be-superseded PUT. The key "can't be" wrong (seg+offset came from the index). Corrupt length: only used to consider/"unconsider" segments for compaction. Worst case: Segment is compacted when it isn't "ripe" yet, or it's not even though it would be.

Acceptable.

2. The main loop in compact_segments. If the key of a live entry is corrupted it could be silently skipped.

Not acceptable

D: Still use this for superseding entries, don't use for the main loop in compact_segments. For something like "borg delete" or "borg prune" it still significantly reduces the reads done.